### PR TITLE
Use threads, cache Scala compiler in GH mvn workflow

### PR DESCRIPTION
**Pull request by gerashegalov**
- Save the overhead of setting up the runner by folding the short-lived runner get-shim-versions-from-dist steps into cache-dependencies job
-  Add  restore-keys to make sure  cache-dependencies does not start on an empty m2, stable versions like for spark artifacts do not change
- Use de.qaware.maven:go-offline-maven-plugin suggested by @revans2 
- Save scala compiler bridge into a cacheable location 

Previous mvn workflow duration: 
- before #9674: ~8min or much longer with retries
- on cache hit: ~7min
- on cache miss: ~11min 

After this PR:
- on cache hit: 5min
- on cache miss: 7min

### CODE REVIEW

**Review by YanxuanLiu — DISMISSED** [review]
LGTM

File: integration_tests/run_pyspark_from_build.sh [hunk 1]
```diff
@@ -197,11 +197,12 @@ else
       "$LOCAL_ROOTDIR"/src/main/python)
 
     REPORT_CHARS=${REPORT_CHARS:="fE"} # default as (f)ailed, (E)rror
+    PYTEST_COLOR=${PYTEST_COLOR:-yes}
```
Comment (jlowe) [inline integration_tests/run_pyspark_from_build.sh:200]: Nit: Change seems unrelated unless there was an error of omission and somewhere else was supposed to override this environment variable to turn of colors?

File: jdk-profiles/pom.xml [hunk 2]
```diff
@@ -43,5 +43,33 @@
...
+        <!-- 
+            TODO does not really belong here but this is a pom without activatedByDefault profiles 
+            that would be deactivated by mistake
+        -->
+        <profile>
+            <id>buildFaster</id>
+            <properties>
+                <maven.scalastyle.skip>true</maven.scalastyle.skip>
+                <maven.scaladoc.skip>true</maven.scaladoc.skip>
+                <rat.skip>true</rat.skip>
```
Comment (jlowe) [inline jdk-profiles/pom.xml:55]: Do all the github workflow builds now build this way, and are we relying on the blossom build to catch scalastyle, scaladoc, and licensing issues?

Comment (gerashegalov) [reply jdk-profiles/pom.xml:55]: it turns out this is not working at all as intended right now, so I am going to table this change, and resort to using just -D properties

File: integration_tests/run_pyspark_from_build.sh (hunk 1 above)
Comment (gerashegalov) [reply integration_tests/run_pyspark_from_build.sh:200]: it should not change anything for existing uses because "yes" is currently hard-coded but I agree it is not related to the main intent of this PR

File: .github/workflows/mvn-verify-check.yml [hunk 3]
```diff
@@ -52,42 +63,34 @@ jobs:
...
           while true; do
-            for pom in pom.xml scala2.13/pom.xml; do 
-              mvn -f $pom help:evaluate -pl dist -Dexpression=included_buildvers \
-              -DforceStdout -PnoSnapshots -q | tr -d ',' | \
-              xargs -n 1 bash -c 'mvn -f $1 initialize -pl sql-plugin-api -am -Dbuildver=$2' _ $pom;
+            for pom in pom.xml scala2.13/pom.xml
+            do
+              # compile base versions to cache scala compiler and compiler bridge
+              mvn ${{ env.COMMON_MVN_FLAGS }} --file $pom \
+                process-test-resources -pl sql-plugin-api -am
```
Comment (jlowe) [inline .github/workflows/mvn-verify-check.yml:77]: Should this be part of the for loop just after this (with the same loop setup) so it can benefit from retry logic?  It seems to defeat some of the dependency resolving since this will start to download dependencies, at least of the sql-plugin-api project.

Comment (gerashegalov) [reply .github/workflows/mvn-verify-check.yml]: done @jlowe

**Review by jlowe — APPROVED** [review]

**Comment by gerashegalov** [comment]
build

**Review by NvTimLiu — APPROVED** [review]
LGTM
